API Add GitHubMarkdownService and update composer dependencies

robbieaverill opened the pull request:
* Replace dflydev/markdown with GitHub markdown parsing via API
* Tidy up readme formatting
* Fix some composer validation failures in composer.json
* Change editor config to be closer to PSR-2, add doc blocks to BuildTasks

This resolves #141, resolves #61, resolves #90 and resolves #115.

# To-do

- [x] Use GitHub API to render markdown instead of erusev/parsedown
- [x] Update guzzlehttp/guzzle for general use
- [x] Ensure legacy guzzle/guzzle still works with Packagist API
- [x] Test update/build addon tasks and frontend

robbieaverill commented:
- [ ] Use GitHub API to render markdown instead of erusev/parsedown
- [ ] Update guzzlehttp/guzzle for general use
- [ ] Ensure legacy guzzle/guzzle still works with Packagist API

robbieaverill commented:
- [ ] I'll do a bit of build/update testing for the addons before this gets reviewed or merged.

dhensby commented on mysite/code/tasks/DeleteRedundantAddonsTask.php:
```diff
@@ -1,37 +1,53 @@
+                $addon->Screenshots()->removeAll();
+                $addon->CompatibleVersions()->removeAll();
+
+                foreach ($addon->Versions() as $version) { /** @var AddonVersion $version */
```
Inline phpDoc should be on line above

dhensby commented on mysite/code/tasks/DeleteRedundantAddonsTask.php:
```diff
@@ -1,37 +1,53 @@
+
+        foreach ($addons as $addon) { /** @var Addon $addon */
+            try {
+                $addon->Keywords()->removeAll();
```
Shouldn't all the relation unmatching be done `onBeforeDelete` on `Addon`?

robbieaverill replied:
I hadn't actually looked much at these parts, it's mostly showing these lines up because of indentation changes

dhensby commented on mysite/code/tasks/BuildAddonsTask.php:
```diff
@@ -1,42 +1,66 @@
-		if($request->getVar('addons')) {
-			$addons = $addons->filter('Name', explode(',', $request->getVar('addons')));
-		}
+        foreach($addons as $addon) { /** @var Addon $addon */
```
Inline type hint to go on line above, please

dhensby commented on mysite/code/services/AddonBuilder.php:
```diff
@@ -91,8 +90,8 @@ private function buildReadme(Addon $addon, $path) {
-				$parser = new MarkdownParser();
-				$readme = $parser->transformMarkdown(file_get_contents($path));
+				$parser = GitHubMarkdownService::create();
+				$readme = $parser->toHtml(file_get_contents($path));
```
if the parser fails (API timeout, rate limit, GH down, whatever) then we're going to override all the readmes with empty string.

Can we attempt to catch that and no override the readme?

robbieaverill commented:
Hey @dhensby - I've moved the PHPDoc lines and added a handler for where the markdown parser fails or returns nothing.

As for the context - there isn't a way to pass a repository context to the GitHub markdown API endpoint. The relative URLs and paths are handled in #143

dhensby commented:
> As for the context - there isn't a way to pass a repository context to the GitHub markdown API endpoint.

There is: https://developer.github.com/v3/markdown/#parameters

robbieaverill commented:
Heh true. So this would give us auto linked issues, pull requests and user mentions. I notice that the docs seem to imply that GitHub themselves don't use context aware parsing for readme files.

Should be a quick change to make anyway though, so I'll throw it in shortly.

dhensby commented:
> I notice that the docs seem to imply that GitHub themselves don't use context aware parsing for readme files.

Yes, though this surely isn't right as I've seen readmes reference issues by number before... unless they'd just hard coded that markdown...

Regardless - using an external API and the risks that brings should provide some tangible benefit over using a library to out-weigh the risks of it being unreliable (ie: sometimes down)

robbieaverill commented:
@dhensby: cool, changes pushed.

For the sake brevity, here are my thoughts overall for using the GitHub API

## Benefits

* Consistency between a repository's readme in GitHub and addons.ss (primary reason for doing this)
* Ability to render GitHub flavoured markdown
* Fix relative links and images (covered in #143) - not directly a cause of having this implemented though
* Don't rely on third party libraries like michelf/php-markdown or erusev/parsedown which erode and get abandoned over time (php-markdown is a good example - it continues to become less and less friendly with modern web apps)
* GitHub can control how markdown is rendered. If they decide to parse relative links to full links in future, we don't need to change anything - yay!

## Drawbacks

* API call while building every module: this is offset by the fact that we build behind the scenes. I wouldn't advocate implementing this if we were rendering on a page load, say
* Riskier: networks can die
* Slower: cross-network calls are slower than loading an installed library

dhensby commented on mysite/code/services/GitHubMarkdownService.php:
```diff
@@ -107,4 +134,22 @@ public function getHeaders()
+            array(
+                'text'    => $markdown,
+                'mode'    => 'gfm',
+                'context' => $this->getContext()
```
Does this work if there's no context?

dhensby commented on mysite/code/services/AddonBuilder.php:
```diff
@@ -91,7 +91,9 @@ private function buildReadme(Addon $addon, $path) {
                 }
 
                 $parser = GitHubMarkdownService::create();
+                $parser->setContext($addon->Name);
```
I think we need to be a bit more clever here. The context is the the `[user]/[repo]` string not just the name of the add on (which could be different) - for example `silverstripe/fulltextsearch` is actually `silverstripe/silverstripe-fulltextsearch` on github.

I think we should check the repo URL is on GH and also determine the repo slug.

If it's not a GH repo  then don't set the context. If the context isn't set the MarkdownService should just use the RAW endpoint (or not set the context if that's the same result)

robbieaverill commented:
I'll be coming back to this

robbieaverill commented:
@dhensby - sorry, took me a while to get back to this but I've updated as per your comments:

* Check for a GitHub context in the repo URL
* Get the "context" from the URL, not the composer package name
* Parse as "raw" mode if not GitHub context, and "gfm" mode if GitHub context
* Added our first unit test! 🎉

dhensby commented:
Fantastic

dhensby commented:
I guess my final question is... Have you tested this locally to ensure it works on our dataset?

Are there API restrictions, will we need an API ket?

robbieaverill commented:
Yeah it will probably need a key to prevent rate limiting. I will add this
into a separate PR if that suits you
On Wed, 11 Jan 2017 at 4:57 AM, Daniel Hensby <notifications@github.com>
wrote:

> I guess my final question is... Have you tested this locally to ensure it
> works on our dataset?
>
>
> Are there API restrictions, will we need an API ket?
>
>
>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/silverstripe/addons.silverstripe.org/pull/142#issuecomment-271614168>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE7lnk3_582ALT66YsQbWW3Ne-5RLtAaks5rQ6p4gaJpZM4LGL5F>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

dhensby commented:
Better use an environment var and keep it out of code. But it will need config somehow

robbieaverill commented:
Yeah totally. I think it's a couple of headers for the guzzle req
